circom/aes256.go

```diff
@@ -0,0 +1,311 @@
+}
+
+// required for galoisMul
+func (aes *AES256) shiftLeft(a uints.U8, size, shift int) uints.U8 {
+	uapi, err := uints.New[uints.U32](aes.api)
+	if err != nil {
+		return uints.U8{}
+	}
+	bits := aes.api.ToBinary(a.Val, size)
+	x := make([]frontend.Variable, size)
+	for i := 0; i < size; i++ {
+		if i < shift {
+			x[i] = 0
+		} else {
+			x[i] = bits[i-shift]
+		}
+	}
+	bytes := aes.api.FromBinary(x...)
+	return uapi.ByteValueOf(bytes)
+}
+
+// shifts state matrix rows
+func (aes *AES256) shiftRows(state [16]uints.U8) [16]uints.U8 {
+	var newState [16]uints.U8
+	for i := 0; i < 4; i++ {
+		newState[i] = state[i] // 0, 1, 2, 3 == t0
+	}
+	for i := 0; i < 4; i++ {
+		newState[4+i] = state[4+((i+1)%4)] // 1, 2, 3, 0 == t1
+	}
+	for i := 0; i < 4; i++ {
+		newState[8+i] = state[8+((i+2)%4)] // 2, 3, 0, 1 == t2
+	}
+	for i := 0; i < 4; i++ {
+		newState[12+i] = state[12+((i+3)%4)] // 3, 0, 1, 2 == t3
+	}
+	return newState
+}
+
+// adds xor and shifts bytes in matrix to match next round representation requirements
+func (aes *AES256) addRoundKey(state [16]uints.U8, expandedKey [240]uints.U8, from int) [16]uints.U8 {
+	var newState [16]uints.U8
+	for i := 0; i < 4; i++ {
+		newState[i] = aes.variableXor(state[i], expandedKey[from+(4*i)], 8)
+		newState[4+i] = aes.variableXor(state[4+i], expandedKey[from+(4*i)+1], 8)
+		newState[8+i] = aes.variableXor(state[8+i], expandedKey[from+(4*i)+2], 8)
+		newState[12+i] = aes.variableXor(state[12+i], expandedKey[from+(4*i)+3], 8)
+	}
+	return newState
+}
+
+// different re-arrangement of variables
+func (aes *AES256) addRoundKey2(state [16]uints.U8, expandedKey [240]uints.U8, from int) [16]uints.U8 {
+	var newState [16]uints.U8
+	ctr := 0
+	for i := 0; i < 4; i++ {
+		newState[i] = aes.variableXor(state[i], expandedKey[from+ctr], 8)
+		newState[4+i] = aes.variableXor(state[4+i], expandedKey[from+ctr+1], 8)
+		newState[8+i] = aes.variableXor(state[8+i], expandedKey[from+ctr+2], 8)
+		newState[12+i] = aes.variableXor(state[12+i], expandedKey[from+ctr+3], 8)
+		ctr += 4
+	}
+	return newState
+}
+
+// xor on bits of two frontend.Variables
+func (aes *AES256) variableXor(a uints.U8, b uints.U8, size int) uints.U8 {
+	uapi, err := uints.New[uints.U32](aes.api)
+	if err != nil {
+		return uints.U8{}
+	}
+	bitsA := aes.api.ToBinary(a.Val, size)
+	bitsB := aes.api.ToBinary(b.Val, size)
+	x := make([]frontend.Variable, size)
+	for i := 0; i < size; i++ {
+		x[i] = aes.api.Xor(bitsA[i], bitsB[i])
+	}
+	bytes := aes.api.FromBinary(x...)
+	return uapi.ByteValueOf(bytes)
+}
+
+// expands 32 byte key to 240 byte output
+func (aes *AES256) expandKey(key [32]uints.U8, sbox0 []uints.U8, RCon []uints.U8) [240]uints.U8 {
+
+	var expand [240]uints.U8
+	i := 0
+
+	for i < 32 {
+		expand[i] = key[i]
+		expand[i+1] = key[i+1]
+		expand[i+2] = key[i+2]
+		expand[i+3] = key[i+3]
+
+		i += 4
+	}
+
+	for i < 240 {
+		t0 := expand[i-4]
+		t1 := expand[i-3]
+		t2 := expand[i-2]
+		t3 := expand[i-1]
+
+		if i%32 == 0 {
```

Comment:
This is the point I worry here on the boundaries

Reply:
I don't quite understand what the boundary here is? The expandkey here is what is needed for each round of confusion to generate the 14 rounds of aes256 encryption. What's needed for each round of obfuscation is exactly 16 bytes. All calculated internally